sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/udaf.scala

```diff
@@ -450,3 +452,165 @@ case class ScalaUDAF(
 
   override def nodeName: String = udaf.getClass.getSimpleName
 }
+
+/**
+ * The internal wrapper used to hook a [[UserDefinedImperativeAggregator]] `udia` in the
+ * internal aggregation code path.
+ */
+case class ScalaUDIA[T](
+    children: Seq[Expression],
+    udia: UserDefinedImperativeAggregator[T],
+    mutableAggBufferOffset: Int = 0,
+    inputAggBufferOffset: Int = 0)
+  extends TypedImperativeAggregate[T]
+  with NonSQLExpression
+  with UserDefinedExpression
+  with ImplicitCastInputTypes
+  with Logging {
+
+  def dataType: DataType = udia.resultType
+
+  val inputTypes: Seq[DataType] = udia.inputSchema.map(_.dataType)
+
+  def nullable: Boolean = true
```

Comment:
Should this also be delegated to the UDIA?

Reply:
I took that default from the `ScalaUDAF` class above, which I tried to keep consistent with wherever it made sense. Possibly it could be exposed to UDIA as well. For example my custom aggregators would be empty (aka identity) as opposed to null, even on null inputs.

Reply:
I think this is going to depend on the function implementation and should be delegated.